### ISSUE TITLE
Reliability of student metrics data by condition

### ISSUE THREAD

**Issue by amurphy-cl** [CarnegieLearningWeb/UpGrade#94]
In the metrics monitor, students who are assigned a condition show up in the data. However, students _not_ assigned a condition are also showing up in the data (for example, if there is a typo in the condition name). This is not the right behavior, and students who are assigned an invalid condition should not appear in the metrics log.

**Comment by nirmalpatel** [comment]
@amurphy-cl can you give more details about how to reproduce this bug? I'm not sure how condition name typo error would happen (unless someone was manually doing the markExpPoint())

**Comment by jerith666** [comment]
The typo isn't in the code that calls `markExperimentPoint()`.  It's in the definition of the experiment (because we don't have validation, cf #26).

The application does call `failedExperimentPoint()` instead of `markExperimentPoint()` in this case, but it appears UpGrade is not taking note of that & using it to exclude metrics about that user.

**Comment by nirmalpatel** [comment]
@msandbothe @jerith666 should we add a CONDITIONS field in the new context metadata that only allows selecting valid conditions when creating experiments?

**Comment by jerith666** [comment]
Aren't we already heading towards that in #114?

**Comment by nirmalpatel** [comment]
We are, but we haven't added the CONDITIONS constraint yet.

Let us do it in #114 if you think it is easy to make a list of possible conditions. Can leave it for later if the condition name list is difficult to make for now.

**Comment by jerith666** [comment]
Ah, yes, I see.  Yeah, it's not too hard to create a list of conditions under each element in `EXP_IDS` (which in our case is a list of variants under each section id).

**Comment by nirmalpatel** [comment]
@jerith666 is it okay if we have a `CONDITIONS` field that is independent of the `EXP_IDS` field? Right now in the UI, the user first enters the condition names and then enters the experiment IDs. To make the nesting work where condition names are within experiment IDs, we will have to ask the user to enter the ID first and then enter the condition name.

We can make simple condition filtering work in the first go, and then go for nesting.

What do you think?

**Comment by jerith666** [comment]
Yeah, I think it's okay to keep them orthogonal to start with.  That'll still help a lot with avoiding typos.

**Comment by amurphy-cl** [comment]
We will work on addressing this in the context metadata design

**Comment by amurphy-cl** [comment]
We will finalize the context metadata design for summer 2022.

**Comment by SritterCL** [comment]
This bug should focus on the behavior of the metrics monitor. The monitor should reflect the students who were actually marked in a particular condition. If the experiment on the UpGrade side defines conditions A and B and data comes in for a student marked in condition C, that should *not* be reflected in the experiment metrics. Similarly, if there is some failure such that a student is assigned to condition A but the student didn't actually get marked with that condition, then that student's data should not be included in the metrics data for condition A. Issue #275 provides the data to UpGrade necessary to implement this.
